### PR TITLE
(feat) - prevent json injection

### PR DESCRIPTION
**Pull request by JoviDeCroock**
Some tests are now skipped since functions won't pass through the new `_` assignment. I opted for this above __proto__ since proto is in essence deprecated and would probably require a lot of changing in how we work in /debug.

This adds a property on vnode that tracks itself, noone can insert an object with the same ref so this will prevent JSON Injection.

TEST2: __proto__ overrides component.constructor implying that catchErrorInComponent stops working. I could really be approaching this wrong but this is what I conclude after testing it with __proto__ instead of _ with obj ref.

ADDS: 12B to core

### CODE REVIEW

File: test/browser/render.test.js [hunk 1]
```diff
@@ -173,7 +179,7 @@ describe('render()', () => {
 		expect(scratch.innerHTML).to.equal('');
 	});
 
-	it('should throw an error on function children', () => {
+	it.skip('should throw an error on function children', () => {
```
Comment (JoviDeCroock) [inline test/browser/render.test.js]: function children don't seem to pass createElement making them victim to the _ check

File: test/browser/toChildArray.test.js [hunk 2]
```diff
@@ -59,7 +59,7 @@ describe('props.children', () => {
 		expect(scratch.innerHTML).to.equal('<div></div>');
 	});
 
-	it('throws an error if given a function child', () => {
+	it.skip('throws an error if given a function child', () => {
```
Comment (JoviDeCroock) [inline test/browser/toChildArray.test.js]: Same story here

File: src/diff/index.js [hunk 3]
```diff
@@ -31,6 +30,11 @@ export function diff(dom, parentDom, newVNode, oldVNode, context, isSvg, excessD
 		oldVNode = EMPTY_OBJ;
 	}
 
+	// When passing through createElement it assignes the object
```
Comment (robertknight) [inline src/diff/index.js]:
```suggestion
	// When passing through createElement it assigns the object
```

File: src/index.d.ts [hunk 4]
```diff
@@ -28,6 +28,7 @@ declare namespace preact {
 		 * Default value: `-1`
 		 */
 		endTime?: number;
+		_: VNode;
```
Comment (robertknight) [inline src/index.d.ts]: No other private properties are documented here. Is there a reason to document this one?

File: test/browser/render.test.js (hunk 1 above)
Comment (robertknight) [reply test/browser/render.test.js]: How does React handle this case?

File: src/create-element.js [hunk 5]
```diff
@@ -62,6 +62,7 @@ export function createVNode(type, props, text, key, ref) {
 		_lastDomChild: null,
 		_component: null
 	};
+	vnode._ = vnode;
```
Comment (robertknight) [inline src/create-element.js]: Could this be given a more self-documenting name, which is then shortened in the build process like the other private properties are?

File: test/browser/render.test.js (hunk 1 above)
Comment (JoviDeCroock) [reply test/browser/render.test.js]: Well are functions as children allowed? Normally it should throw and now they won't get passed. Don't really know what we should do

File: src/create-element.js (hunk 5 above)
Comment (JoviDeCroock) [reply src/create-element.js]: Is this better, it's _self now. I must admit the name was a bit odd of me to do, I wanted to contain bundle size increase but you're right that's what we have our mangle file for :D thanks for pointing it out

File: test/browser/render.test.js (hunk 1 above)
Comment (robertknight) [reply test/browser/render.test.js]: React produces this detailed error if you try to do this:

Screenshot 2019-04-14 23 13 57

This logic could be moved to the `vnode` validation hook in `preact/debug` perhaps? The current behaviour on this branch of rendering nothing if a function child is passed to a DOM node seems reasonable to me, although that should have a test if it is agreed upon.

Comment (JoviDeCroock) [reply test/browser/render.test.js]: Is this in scope of this pr?

**Comment by JoviDeCroock** [comment]
I'll make a seperate PR one of these days to add it in debug

File: test/browser/toChildArray.test.js (hunk 2 above)
Comment (robertknight) [reply test/browser/toChildArray.test.js]: Does this test cover anything the other one to do with function children doesn't? If not I would suggest to just remove it.

**Comment by JoviDeCroock** [comment]
> I have a query over whether there is any reason to keep the remaining skipped test. Overall LGTM though.

Good point, I must've looked over it apologies for that. Will fix it asap

**Pull request merged**